### ISSUE TITLE
corrupt channel behavior, and non-deterministic interaction with other, unrelated channels

### ISSUE THREAD

**Issue by antoniofabio** [nextflow-io/nextflow#2111]
## Bug report 

It was hard to come up with a good title. I've got a script which shows erratic behavior, where the results from one channel are sometimes (more often than not) seemingly corrupted by the presence of actions in another, unrelated channel.

### Expected behavior and actual behavior
Given the following script:
```
nextflow.enable.dsl=2

workflow {
    samples = Channel.from(['prefix/s1/chr1'])
    refs = Channel.from(['/a/prefix/chr5'])

//<--- adding/removing this block makes all the difference
    chrID_sampleID_geno =
      samples |
      map {
        fields = it.toString().split('/')
        chromosome_id = fields[ fields.size() - 1 ]
        sample_id = fields[ fields.size() - 2 ]
      }
//--->

    refs |
      map {
          fields = it.toString().split('/')
          chromosome_id = fields[ fields.size() - 1 ]
          [ chromosome_id, it ]
      } |
      view()
}
```
I expect to see printed in output:
`
[chr5, /a/prefix/chr5]
`

But if I run it multiple times, sometimes I see this instead:
`
[chr1, /a/prefix/chr5]
`
### Environment 

* Nextflow version: 20.10.0 build 5430
* Java version: 1.8.0_211
* Operating system: Linux
* Bash version: 4.2.46(2)-release

### Additional context

I'm probably doing something silly in my script. Still, the end result looks wild

**Comment by GDCCP** [comment]
I haven't tried running your code as I don't have 20.10.0 ready. However, try to add 'def' for variables used in the closure(fields,chromosome_id in particular) like:

def fields = it.toString.split('/')

**Comment by antoniofabio** [comment]
Adding 'def's everywhere does the trick, in that it makes the problem go away.
It seems indeed there's some scoping issue at play. Having no groovy experience, I guess I unknowingly relied on too much magic happening?

Btw, the problem is devilish, in that one can (_occasionally_) get plain wrong results from innocently looking bits of code, without a single hint of an error nor a warning.

**Comment by GDCCP** [comment]
have a look at issue #678. I am not sure if the current nextflow 'best practices' is documented.

I view the 'nextflow run' as a 2 phase execution -- parsing the .nf files and spawning jobs. The nextflow processes/workflows are some kind of object templates that are created in phase 1. These objects are instantiated into concrete jobs(OS processes) in phase 2 with the values from the channels. A nextflow process can have multiple live instances and I have to avoid these instances to share a groovy variable -- so I need to be defensive.

All my variables are local (with 'def') and I am using many module files. Besides using addParams to pass values around, I take advantage of the params/workflow global objects. These values are set in the first phase and then treated as read-only after initialization with no exception. You can guess I am using functional programming practices heavily.

**Comment by antoniofabio** [comment]
Thanks GDCCP, though your last comment mostly flew way over my head. It seems I need a deeper understanding of groovy and how the system works in getting from the .nf files to the final workflow results.

So far I assumed that implicitly declared variables (i.e. without 'def', which I didn't know existed) were lexically scoped, and I was just wrong. Then how this dynamic/global scoping interacts with the nextflow DSL layer is an entirely separate issue...

I'll try to code more defensively going forward, and probably just give up having a solid grasp of how groovy+nextflow dsl actually work. What could go wrong? :D
